### ISSUE TITLE
Quartznet: replicating Training

### ISSUE THREAD

**Issue by Giuseppe5** [NVIDIA/NeMo#230]
Hi,
I would like to replicate the training process of quartznet, however I'm having some issues.

In particular, on the official paper and on [website](https://ngc.nvidia.com/catalog/models/nvidia:quartznet15x5) it says that a speed perturbation was applied to the dataset (±10%). However I can't seem to find trace of that neither in the download script, nor in the code.

The other issue I'm having is related to the hyperparameters used when training on Librispeech, such as learning rate and warmup steps. They are not mentioned explicitly anywhere, so I was wondering if you could help me with that as well.

Thank you!

**Comment by okuchaiev** [comment]
>> "In particular, on the official paper and on website it says that a speed perturbation was applied to the dataset (±10%)."
This was done by outside script once - the full LibriSpeech train data is effectively tripled by it so that for training we used: LibrisSpeech + LibriSpeech-slowed10% + LibriSpeech-increased10%.
@blisc can give more details (or maybe even script) on how that was done. But this script isn't a part of NeMo.

>> "related to the hyperparameters used when training on Librispeech, such as learning rate and warmup steps."
- This depends on your training setup a lot. Expect ~week of training on DGX-1V and ~4 hours if you have a cluster of 32 DGX2s. So the training is computationally heavy which is why we also provide pre-trained checkpoints.
Can you share your training setup details? Then we can recommend hyperparameters. What we are interested in is the number of workers (e.g. GPUs) and their memory size

**Comment by Giuseppe5** [comment]
Thanks for the clarification on the speed perturbation. I noticed that on the official Jasper implementation they use the same approach, and then in the manifest and dataset classes they handle the presence of the same audio file with three different speeds . In Quartznet there seems to be no explicit handling of this, otherwise I would have tried to reuse and reimplement their script. 

Regarding the hyperparameters, I would be interested in the configuration for a DGX-1V.
Thank you!

**Comment by okuchaiev** [comment]
To be honest, I don't remember exact hyperparams, but they were something like this:
1) LR policy - I think we used SquareAnnealing in this particular case but now I think CosineAnnealing works better. Starting LR was 0.015 (pretty sure about that)
2) Optimization algorithm: Novograd, I think we used beta2=0.0 but you might want to set it to 0.25 or 0.5
3) Batch size was 64 per GPU (will fit on 32GB card, but not 16GB)
4) Number of epochs were ~ 135
5) quartznet15x5.yaml - no changes there
6) weight_decay was either 0.001 or 0.0001
Also, keep in mind that for Common Voice we used full "validated", also CV dataset is continuously improving.
Sorry, I don't have exact params - I'll try to reproduce/improve the model and will post it on documentation.

We do, however, have the best hyperparams for the best Quartznet15x5 model from this paper: https://arxiv.org/abs/1910.10261  But this model was trained without MCV, only on LS + LS speed perturbed.

They are:
1) batch_size=32
2) num_epochs=134
3) cosine LR policy (and novograd as optimizer)
4) beta1=0.95, beta2=0.5
5) warmup=1000
6) lr=0.01
7) weight_decay=0.001
Trained on 1 DGX-1V with 32GB cards and O0 optimization flag

**Comment by Giuseppe5** [comment]
That's very helpful, thank you! 
Is there any chance that you will release the pretrained model used in the paper for LS-LS speed perturbed only?

**Comment by okuchaiev** [comment]
@redoctopus could you please put this model (from paper) on NGC and ask @pchitale1 to make it public? Thanks!

**Comment by redoctopus** [comment]
Sure, on it!

**Comment by Giuseppe5** [comment]
While waiting for the pretrained model of the paper, I was wondering whether I can take inspiration from Jasper to implement the speed perturbation part.

What they do is, after parsing the dataset, they create the slower and faster version of every track, so each track has 3 version at three speeds (0.9 , 1.0, 1.1).
Then, at training time, for each track they pick randomly one of the three version through some small modifications in the Manifest and Dataset classes.

Could it make sense?

**Comment by blisc** [comment]
Regarding speech perturbation, after creating the slower, and faster version, you can just add these files to your manifest. It will end up looking something like:
```
...
{"audio_filepath": "<absolute_path_to>/1355-39947-0000.wav", "duration": 11.3, "text": "psychotherapy and the community both the physician and the patient find their place in the community the life interests of which are superior to the interests of the individual"}
{"audio_filepath": "<absolute_path_to>/1355-39947-0000-sp-0p9.wav", "duration": 12.43, "text": "psychotherapy and the community both the physician and the patient find their place in the community the life interests of which are superior to the interests of the individual"}
{"audio_filepath": "<absolute_path_to>/1355-39947-0000-sp-1p1.wav", "duration": 10.17, "text": "psychotherapy and the community both the physician and the patient find their place in the community the life interests of which are superior to the interests of the individual"}
...
```
and adjust num_epochs accordingly. For example, if your non-speed perturbed run is 400 epochs, use 134 epochs for your new run using speed perturbed data.

**Comment by Oktai15** [comment]
@Giuseppe5 were you able to reproduce training?

**Comment by Giuseppe5** [comment]
@blisc Thank you Jason for the clarification! I prepared everything for the speed perturbation and I will try to fine tune the model with the pretrained version that has recently been uploaded.
@Oktai15 My goal is just to fine tune the model and not to train it from scratch (although I can make good use of the "original" hyper-parameters), so I don't know if my results could be helpful to you.

**Comment by redoctopus** [comment]
Apologies for the delay--the pretrained QuartzNet model (trained using only LibriSpeech w/ speed perturbation) is up on NGC now: https://ngc.nvidia.com/catalog/models/nvidia:quartznet_15x5_ls_sp

**Comment by tugstugi** [comment]
@okuchaiev what was the min learning rate for the cosine scheduler for only LibriSpreech trained model?

**Comment by ArtemisZGL** [comment]
@okuchaiev hello, have you tried the O1 flag for mixed precision training ? I found if I use O1 the result was quiet bad, but the O0 result is normal with same other param.

**Comment by bagavi** [comment]
@Giuseppe5 : What are you fine tunning it for? What dataset and parameters did you use? Thanks!
